### PR TITLE
Add snippet of API for integration with Catalyst.

### PR DESCRIPTION
**Pull request by VladSkripniuk**
This PR addresses integration with Catalyst proposed in #1055

### CODE REVIEW

**Comment by hvy** [comment]
I understand this is related to GSoC 2020. @VladSkripniuk do you want us to start reviewing this? I'm wondering partially because of GSoC but also because the PR is missing tests and there are some minor CI failures. It wasn't at a first glance clear to me what the status of this PR was. Thanks!

**Comment by VladSkripniuk** [comment]
@hvy thank you for getting back to me! This PR is not yet ready for review, I'll implement `CatalystPruningCallback` and add tests in the following days. I'll add [WIP] to the title of PR.

**Comment by hvy** [comment]
Thanks for the quick response. Got it.

**Comment by VladSkripniuk** [comment]
@hvy could you please review this PR? I implemented the pruning callback and fixed CI failures.

**Comment by hvy** [comment]
Sorry for the delayed response. It's mainly because I'm not too familiar with Catalyst, but also because of the WIP in the title. Is it the case that it's ready for review?

**Comment by VladSkripniuk** [comment]
This PR is ready for review. I'll remove WIP from the title.

File: tests/integration_tests/test_catalyst.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import pytest
+import sys
+
+import torch
+
+if type_checking.TYPE_CHECKING:
+    from typing import Dict  # NOQA
+    from typing import List  # NOQA
+    from typing import Union  # NOQA
```
Comment (smly) [inline tests/integration_tests/test_catalyst.py:19]: [MUST] `typing.*` are not used.

```suggestion
```

File: tests/integration_tests/test_catalyst.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+try:
+    from catalyst.dl import SupervisedRunner
+except ImportError:
+    pass
```
Comment (smly) [inline tests/integration_tests/test_catalyst.py:4]: Question: Is there any reason to use the `try-except` statement here?

File: optuna/integration/catalyst.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import optuna
+
+
+if optuna.type_checking.TYPE_CHECKING:
+    from typing import Any  # NOQA
```
Comment (smly) [inline optuna/integration/catalyst.py:5]: [MUST] type_checking is not necessary in this case. The motivation for the use of type_checking.TYPE_CHECKING constraint is avoiding a circular module dependency in this repo.

There are similar codes in Optuna. For example https://github.com/optuna/optuna/blob/master/optuna/integration/lightgbm_tuner/alias.py#L1-L6. This code is also inappropriate. Sorry for confusing you.

```suggestion
from typing import Any  # NOQA
```

File: examples/catalyst_simple.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
+"""
+Optuna example that optimizes multi-layer perceptrons using Catalyst.
+
+In this example, we optimize the validation accuracy of hand-written digit recognition using
+Catalyst, and MNIST. We optimize the neural network architecture. As it is too time
+consuming to use the whole MNIST dataset, we here use a small subset of it.
```
[the remark this answers is not in the document]
Comment (VladSkripniuk) [reply examples/catalyst_simple.py]: Good catch, I've edited the docstring accordingly.

File: tests/integration_tests/test_catalyst.py (hunk 2 above)
Comment (VladSkripniuk) [reply tests/integration_tests/test_catalyst.py]: As I think about it now, I can't see a reason to use the `try-except` statement here. I'll remove it.

File: examples/catalyst_simple.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    def __init__(self, trial):
+        super(Net, self).__init__()
+        self.layers = []
+        self.dropouts = []
+
+        # We optimize the number of layers, hidden units in each layer and dropouts.
+        n_layers = trial.suggest_int("n_layers", 1, 3)
+        dropout = trial.suggest_uniform("dropout", 0.2, 0.5)
+        input_dim = 28 * 28
+        CLASSES = 10
```
[the remark this answers is not in the document]
Comment (VladSkripniuk) [reply examples/catalyst_simple.py]: Done.

File: tests/integration_tests/test_catalyst.py (hunk 2 above)
Comment (VladSkripniuk) [reply tests/integration_tests/test_catalyst.py]: It turns out `try-except` was necessary to skip catalyst tests for python3.5 gracefully, as catalyst is only available for python version > 3.5.

Comment (smly) [reply tests/integration_tests/test_catalyst.py]: Oh, I missed it. Make sense. Your additional patch using [`pytest.importorskip`](https://docs.pytest.org/en/latest/reference.html#pytest-importorskip) is nice. Thanks for your effort in the backward compatibility!

File: tests/integration_tests/test_catalyst.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import optuna
+from optuna.integration import CatalystPruningCallback
+from optuna.testing.integration import DeterministicPruner
+
+import pytest
+import sys
+
+import torch
```
Comment (smly) [inline tests/integration_tests/test_catalyst.py:8]:
```suggestion
import sys

import pytest
import torch

import optuna
from optuna.integration import CatalystPruningCallback
from optuna.testing.integration import DeterministicPruner
```

File: optuna/integration/catalyst.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import optuna
+
+from typing import Any  # NOQA
```
Comment (smly) [inline optuna/integration/catalyst.py:3]:
```suggestion
from typing import Any  # NOQA

import optuna
```

File: examples/catalyst_simple.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+import optuna
+from optuna.integration import CatalystPruningCallback
+
+import torch
+from torch import nn
+from torch.nn import functional as F
+from torch.utils.data import DataLoader
+from torchvision.datasets import MNIST
+from torchvision import transforms
```
Comment (smly) [inline examples/catalyst_simple.py:33]:
```suggestion
import argparse
import os

from catalyst.dl import AccuracyCallback
from catalyst.dl import SupervisedRunner
import torch
from torch import nn
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import MNIST

import optuna
from optuna.integration import CatalystPruningCallback
```

File: optuna/integration/catalyst.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+    _available = True
+except ImportError as e:
+    _import_error = e
+    # CatalystPruningCallback is disabled because Catalyst is not available.
+    _available = False
+    Callback = object
+
+
+class CatalystPruningCallback(Callback):
```
Comment (hvy) [inline optuna/integration/catalyst.py:17]: I'd like to hear inputs from other developers but I'm wondering if we should start this feature as `experimental` to guard ourselves against near future changes, as `catalyst` is new to the Optuna "ecosystem" and we don't really know what kinds of APIs are expected by Catalyst users?

File: tests/integration_tests/test_catalyst.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        criterion = torch.nn.MSELoss()
+        optimizer = torch.optim.Adam(model.parameters())
+
+        runner = catalyst.dl.SupervisedRunner()
+        runner.train(
+            model=model,
+            criterion=criterion,
+            optimizer=optimizer,
+            loaders=loaders,
+            logdir="./logdir",
```
Comment (hvy) [inline tests/integration_tests/test_catalyst.py:38]: Wondering if this test isn't creating files. If so, we might want to make sure it's cleaned up or use some temporary directory.

**Comment by hvy** [comment]
Hi @VladSkripniuk. I don't by any means intend to stress but I see that you have some ongoing PRs and see that this one likely is the one closest to being merged. Would you be interested in continuing working on it?

**Comment by VladSkripniuk** [comment]
Hi @hvy. I have incorporated proposed changes. I was thinking about integration with Alchemy for experiment logging. We can leave other integrations for different PR.

**Comment by hvy** [comment]
Thanks for your quick response. It looks better now but could you fix  https://github.com/optuna/optuna/pull/1056#discussion_r435839233 as well.

**Comment by hvy** [comment]
Ops, you need to add the introduced test to this ignore list for Python 3.5 CI jobs https://github.com/optuna/optuna/blob/master/.circleci/config.yml#L237-L239. There seem to be some style errors are well, could you check that with black? https://app.circleci.com/pipelines/github/optuna/optuna/1338/workflows/718da8b6-8ea1-4b3b-8032-53dfdebfe756/jobs/47161/steps

Again, thank you for your swift action.

**Comment by hvy** [comment]
Could you just add the introduced example to the GitHub Action example build ignore list? https://github.com/optuna/optuna/blob/master/.github/workflows/examples.yml#L39 Otherwise LGTM. Again, place the `experimental` decorator on this class for now. If you are okay, I could do that in a follow up PR.

**Comment by VladSkripniuk** [comment]
I've added catalyst example to build ignore list. Please add the experimental decorator in a follow up PR. Thank you!

**Pull request merged**